README.md
## frp-otp

一个专注于访问安全的 [frp server manager plugin](https://github.com/fatedier/frp/blob/master/doc/server_plugin_zh.md) 实现，掌控每个接入的请求，让你对进入 `frps` 的连接了如指掌，不再裸奔。

## 快速启动

[下载地址](https://github.com/csg2008/frp-otp/releases)

### 目录介绍

```bash
* frp-otp
└─── system
|    |      frp-otp.service                  # linux 系统服务配置文件
|
│           frp-otp                          # frp-otp 程序
|           frp-otp.json                     # 通知插件配置文件
```

### 打印帮助信息

```bash
./frp-otp --h
```

### 命令行启动

```bash
./frp-otp start -c ./frp-otp.json
```

## 配置介绍

### frps

在 `frps.ini` 增加以下配置

```
[plugin.frp-otp]
addr = 127.0.0.1:9000                              // frp-otp 地址
path = /handler                                    // frp-otp url, 固定配置
ops = Login,NewProxy,NewWorkConn,NewUserConn       // 通知的操作
```

### 黑白名单配置（`IP` 过滤）

先判断白名单，后判断黑名单。如果启用强制验证，未配置到通道列表的代理默认是会拒绝访问的，否则不对请求的 IP 进行验证。已经配置到通道列表中的代理，如果运行模式配置为 bypass 默认是充许所有请求，不对请求的 IP 进行验证

```json
{
    "enforce": true,                               // 是否启用强制验证
    "domain": "zero.frp",                          // 域名
    "BindAddress": "127.0.0.1:9000",               // 服务监听的地址
    "Channel": {                                   // 通道配置
        "dev.web": {                               // 通道名
            "status": true,                        // 通道状态
            "model": "bypass",                     // 运行模式
            "secret": "...",                       // OTP 密钥，可以通过 key 命令生成
            "issuer": "dev.web",                   // OTP 组织名
            "notify": null,                        // 访问消息通知，未实现
            "blacklist": [                         // 黑名单
              "127.0.0.1"
            ],
            "whitelist": [                         // 白名单
              "127.0.0.1"
            ],
        },
        ......
    }
}
```
### 接入IP控制
如果通道配置为非 bypass 模式，就需要先访问 http://服务器IP/knock 输入对应基于时间的验证码，只在验证码通过了五分钟内才可以访问代理的通道。一个IP一分钟内只能输错五次验证码，否则就会被加入自动屏蔽的黑名单，且每尝试一次自动延长屏蔽时间五分钟，直到屏蔽时间到期。

CHANGELOG.md
2022-11-30
+ 初始版本

main.go
// 基于 TOTP 的 FRP 访问授权验证
package main

import (
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
)

var appName string
var buildRev string
var buildDate string
var buildVersion string

var entry = flag.String("e", "web", "工具启动入口")
var proxy = flag.String("p", "", "代理通道名")
var configFile = flag.String("c", "frp.otp.json", "OPT 配置文件")
var showVer = flag.Bool("v", false, "显示应用版本信息并退出")
var showHelp = flag.Bool("h", false, "显示应用帮助信息并退出")

// LoadJSON 从文件加载 JSON 到变量
func LoadJSON(file string, in interface{}) error {
	var err error
	var data []byte

	if data, err = ioutil.ReadFile(file); nil == err {
		err = json.Unmarshal(data, in)
	}

	return err
}

// SaveJSON 保存数据到 JSON 文件
func SaveJSON(file string, in interface{}) error {
	var bj, _ = json.Marshal(in)

	return ioutil.WriteFile(file, bj, 0644)
}

// IsFile returns true if given path is a file,
// or returns false when it's a directory or does not exist.
func IsFile(filePath string) bool {
	f, e := os.Stat(filePath)
	if e != nil {
		return false
	}
	return !f.IsDir()
}

// runWeb 运行 WEB 服务器
func runWeb(cfg *Config) error {
	var s, err = NewServer(cfg)
	if nil == err {
		err = s.Run()
	}

	return err
}

// genOtpKey 生成通道KEY
func genOtpKey(cfg *Config, proxy string) error {
	if c, ok := cfg.Channel[proxy]; ok {
		fmt.Println("正在生成通道 " + proxy + " 的 OTP KEY")
		c.GenerateOTP(proxy, cfg.Domain)
	} else {
		if "" == proxy {
			fmt.Println("请输入要生成 OTP KEY 的通道，当前可用通道列表：")
		} else {
			fmt.Println("输入的通道 " + proxy + " 不存在，当前可用通道列表：")
		}

		for k := range cfg.Channel {
			fmt.Println(k)
		}
	}

	return nil
}

func main() {
	flag.Usage = func() {
		fmt.Fprintln(os.Stderr, "使用说明: ", filepath.Base(os.Args[0]), " 参数选项")
		fmt.Fprintln(os.Stderr, "欢迎使用 FRP 访问授权验证工具")
		fmt.Fprintln(os.Stderr, "更多信息: http://github.com/csg800/frp-otp")
		fmt.Fprintln(os.Stderr, "")

		fmt.Fprintln(os.Stderr, "参数选项:")
		flag.PrintDefaults()

		if *showHelp {
			fmt.Fprintln(os.Stderr, "")
			fmt.Fprintln(os.Stderr, "启动入口：")
			fmt.Fprintln(os.Stderr, "          web      启用访问授权验证服务")
			fmt.Fprintln(os.Stderr, "          key      为通道生成访问授权KEY")
		}
	}

	flag.Parse()

	if *showVer {
		fmt.Println(appName + " " + "Ver: " + buildVersion + " build: " + buildDate + " Rev:" + buildRev)
		return
	}

	if *showHelp {
		flag.Usage()
		return
	}

	defer func() {
		if err := recover(); nil != err {
			fmt.Println("system error:", err)
		}
	}()

	var err error
	var cfg = new(Config)
	if IsFile(*configFile) {
		if err = LoadJSON(*configFile, cfg); nil == err {
			if 0 == len(cfg.Channel) {
				err = errors.New("配置文件中通道列表为空")
			} else {
				switch *entry {
				case "web":
					err = runWeb(cfg)
				case "key":
					err = genOtpKey(cfg, *proxy)
				default:
					flag.Usage()
				}
			}
		} else {
			err = errors.New("加载配置文件失败：" + err.Error())
		}
	} else {
		err = errors.New("配置文件 " + *configFile + " 不存在")
	}

	if nil != err {
		fmt.Println("错误：", err)
	}
}


server.go
package main

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"log"
	"net"
	"net/http"
	"os"
	"os/signal"
	"strings"
	"sync"
	"syscall"
	"time"

	plugin "github.com/fatedier/frp/pkg/plugin/server"
)

// Server 服务器
type Server struct {
	ec   int
	cfg  *Config
	s    *http.Server
	h    map[string]func(r *http.Request, req *plugin.Request) *plugin.Response
	done chan struct{}
}

// NewServer 创建服务器
func NewServer(cfg *Config) (*Server, error) {
	s := &Server{
		cfg:  cfg,
		done: make(chan struct{}, 1),
	}
	if err := s.init(); err != nil {
		return nil, err
	}
	return s, nil
}

// Run 开始运行
func (s *Server) Run() error {
	go s.handlerSignal()

	var err = s.s.ListenAndServe()
	<-s.done
	return err
}

// Stop 停止
func (s *Server) Stop() error {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	if err := s.s.Shutdown(ctx); err != nil {
		log.Fatalf("shutdown HTTP server error: %v", err)
	}
	log.Printf("HTTP server exited")
	close(s.done)
	return nil
}

// init 初始化
func (s *Server) init() error {
	s.s = &http.Server{
		Handler: s,
		Addr:    s.cfg.BindAddress,
	}
	s.h = map[string]func(r *http.Request, req *plugin.Request) *plugin.Response{
		plugin.OpPing:        s.handlePing,
		plugin.OpLogin:       s.handleLogin,
		plugin.OpNewUserConn: s.handleNewUserConnect,
		plugin.OpNewProxy:    s.handleNewProxy,
		plugin.OpCloseProxy:  s.handleCloseProxy,
		plugin.OpNewWorkConn: s.handleNewWorkConnect,
	}

	for _, v := range s.cfg.Channel {
		if nil == v.Lock {
			v.Lock = new(sync.RWMutex)
		}
	}

	return nil
}

// handlerSignal 信号处理
func (s *Server) handlerSignal() {
	var sig = make(chan os.Signal, 2)
	var tick = time.After(time.Second * 300)

	signal.Notify(sig, os.Interrupt, os.Kill, syscall.SIGINT, syscall.SIGTERM, syscall.SIGHUP)

	for {
		select {
		case sv := <-sig:
			switch sv {
			case os.Interrupt, syscall.SIGINT:
				if 0 == s.ec {
					log.Println("Send ^C to force exit...")
				}

				if s.ec > 0 {
					s.Stop()
				}

				s.ec++
			case os.Kill, syscall.SIGTERM:
				s.Stop()
			case syscall.SIGHUP:
				s.cfg.GC()
			}
		case <-tick:
			s.cfg.GC()
		}
	}
}

// ServeHTTP WEB 入口
func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	var uri = strings.ToLower(r.URL.Path)

	if "/" == uri {
		s.WebIndex(w, r)
	} else if "/knock" == uri {
		s.WebKnock(w, r)
	} else if "/handler" == uri {
		s.HandleRPC(w, r)
	} else {
		http.NotFound(w, r)
	}
}

// SendJSON 发送JSON输出
func (s *Server) SendJSON(w http.ResponseWriter, r *http.Request, data interface{}) {
	var err error
	var bj []byte

	if nil == data {
		bj = []byte(`{"status":false, "msg":"unknown action"}`)
	} else {
		if bj, err = json.Marshal(data); nil != err {
			bj = []byte(`{"status": false, "msg":"` + err.Error() + `":`)
		}
	}

	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.WriteHeader(http.StatusOK)
	w.Write(bj)
}

// SendHTML 发送 HTML 代码
func (s *Server) SendHTML(w http.ResponseWriter, r *http.Request, status int, html string) {
	w.Header().Set("Content-Type", "text/html;charset=UTF-8")
	w.WriteHeader(status)

	w.Write([]byte(html))
}

// WebIndex 首页
func (s *Server) WebIndex(w http.ResponseWriter, r *http.Request) {
	s.SendHTML(w, r, http.StatusOK, "授权验证")
}

// WebKnock 敲门
func (s *Server) WebKnock(w http.ResponseWriter, r *http.Request) {
	var ip, content string
	var code = r.FormValue("code")
	var channel = r.URL.Query().Get("channel")

	if c, ok := s.cfg.Channel[channel]; ok {
		ip, _, _ = net.SplitHostPort(r.RemoteAddr)
		if ts, block := c.IsBlock(ip); block {
			if ts > 0 {
				c.AddBlock(ip, time.Now().Unix(), 10)
				content = "您的 IP " + ip + " 被自动屏蔽，请于 " + time.Unix(ts, 0).Format("2006-01-02 15:04:05") + " 后再重试"
			} else {
				content = "您的 IP " + ip + " 在黑名单中，请联系管理员"
			}

			http.Error(w, content, http.StatusForbidden)
		} else {
			if "" == code {
				content = "<html><head><title>请输入授权码</title></head><body><p>请输入授权码：</p><form method='POST'><input type='text' id='code' name='code' /><input type='submit' value='提交' /></form></body></html>"
			} else {
				if c.Check(code) {
					c.AddAllow(ip, time.Now().Unix()+300)
					content = "<html><head><title>验证成功</title><body>验证成功，请于五分钟内访问</body></html>"
				} else {
					c.AddBlock(ip, time.Now().Unix(), 10)
					content = "<html><head><title>请输入授权码</title></head><p>授权码错误，请输入授权码：</p><body><form method='POST'><input type='text' id='code' name='code' /><input type='submit' value='提交' /></form></body></html>"
				}
			}

			s.SendHTML(w, r, http.StatusOK, content)
		}
	} else if s.cfg.Enforce {
		http.Error(w, "Forbidden", http.StatusForbidden)
	} else {
		http.Error(w, "Ok", http.StatusOK)
	}
}

// HandleRPC 处理接口过来的请求
func (s *Server) HandleRPC(w http.ResponseWriter, r *http.Request) {
	var err error
	var req *plugin.Request
	var res *plugin.Response

	if req, err = s.parseRequest(r); nil == err {
		if fn, ok := s.h[req.Op]; ok {
			res = fn(r, req)
		} else {
			res = &plugin.Response{
				Reject:       true,
				RejectReason: "unknown operation",
			}
		}
	} else {
		res = &plugin.Response{
			Reject:       true,
			RejectReason: "parse request error: " + err.Error(),
		}
	}

	s.SendJSON(w, r, res)
}

// handlePing 处理心跳相关信息
func (s *Server) handlePing(r *http.Request, req *plugin.Request) *plugin.Response {
	var resp = new(plugin.Response)

	resp.Unchange = true

	return resp
}

// handleLogin 处理用户登录操作信息
func (s *Server) handleLogin(r *http.Request, req *plugin.Request) *plugin.Response {
	var resp = new(plugin.Response)
	// var content = req.Content.(*plugin.LoginContent)
	// var token = content.Metas["token"]

	// if content.User == "" || token == "" {
	// 	resp.Reject = true
	// 	resp.RejectReason = "user or meta token can not be empty"
	// } else if content.User != "" {
	// 	resp.Unchange = true
	// } else {
	// 	resp.Reject = true
	// 	resp.RejectReason = "invalid meta token"
	// }

	resp.Unchange = true

	return resp
}

// handleNewUserConnect 处理创建用户连接 (支持 tcp、stcp、https 和 tcpmux 协议)。
func (s *Server) handleNewUserConnect(r *http.Request, req *plugin.Request) *plugin.Response {
	var ip string
	var resp = new(plugin.Response)
	var content = req.Content.(*plugin.NewUserConnContent)

	if c, ok := s.cfg.Channel[content.ProxyName]; ok {
		if c.Status {
			if "bypass" == c.Model {
				resp.Unchange = true
				resp.Content = content
			} else {
				ip, _, _ = net.SplitHostPort(content.RemoteAddr)
				if c.IsAllow(ip) {
					resp.Unchange = true
					resp.Content = content
				} else {
					resp.Reject = true
					resp.RejectReason = "user ip not in allow Whitelist"
				}
			}
		} else {
			resp.Reject = true
			resp.RejectReason = "proxy is not enable"
		}
	} else if s.cfg.Enforce {
		resp.Reject = true
		resp.RejectReason = "proxy is not config"
	} else {
		resp.Unchange = true
		resp.Content = content
	}

	return resp
}

// handleNewProxy 处理创建代理的相关信息
func (s *Server) handleNewProxy(r *http.Request, req *plugin.Request) *plugin.Response {
	var resp = new(plugin.Response)
	var content = req.Content.(*plugin.NewProxyContent)

	if c, ok := s.cfg.Channel[content.ProxyName]; ok {
		if c.Status {
			resp.Unchange = true
			resp.Content = content
		} else {
			resp.Reject = true
			resp.RejectReason = "proxy is not enable"
		}
	} else if s.cfg.Enforce {
		resp.Reject = true
		resp.RejectReason = "proxy is not config"
	} else {
		resp.Unchange = true
		resp.Content = content
	}

	return resp
}

// handleCloseProxy 处理代理关闭。(仅用于通知)
func (s *Server) handleCloseProxy(r *http.Request, req *plugin.Request) *plugin.Response {
	var resp = new(plugin.Response)

	resp.Unchange = true

	return resp
}

// handleNewWorkConnect 处理创建工作连接
func (s *Server) handleNewWorkConnect(r *http.Request, req *plugin.Request) *plugin.Response {
	var resp = new(plugin.Response)

	resp.Unchange = true

	return resp
}

// parseRequest 解析请求数据
func (s *Server) parseRequest(r *http.Request) (*plugin.Request, error) {
	var err error
	var op string
	var findOp bool
	var data []byte
	var idx, cnt int
	var body interface{}
	var req *plugin.Request

	if data, err = io.ReadAll(r.Body); nil == err {
		for k, v := range data {
			if '"' == v {
				if k > 0 && '\\' == data[k-1] {
					continue
				}

				cnt++
				if 1 == cnt%2 {
					idx = k
				} else {
					if k-idx == 3 && (('o' == data[k-2] || 'O' == data[k-2]) && ('p' == data[k-1] || 'P' == data[k-1])) {
						findOp = true
					} else if findOp {
						op = string(data[idx+1 : k])

						break
					}
				}
			}
		}

		if findOp {
			switch op {
			case plugin.OpPing:
				body = &plugin.PingContent{}
			case plugin.OpLogin:
				body = &plugin.LoginContent{}
			case plugin.OpNewUserConn:
				body = &plugin.NewUserConnContent{}
			case plugin.OpNewProxy:
				body = &plugin.NewProxyContent{}
			case plugin.OpCloseProxy:
				body = &plugin.CloseProxyContent{}
			case plugin.OpNewWorkConn:
				body = &plugin.NewWorkConnContent{}
			default:
				err = errors.New("unknown operation")
			}

			if nil == err {
				req = &plugin.Request{
					Content: body,
				}

				err = json.Unmarshal(data, req)
			}
		}
	}

	return req, err
}


config.go
package main

import (
	"bytes"
	"fmt"
	"image/png"
	"io/ioutil"
	"sync"
	"time"

	"github.com/pquerna/otp"
	"github.com/pquerna/otp/totp"
)

// Config 配置
type Config struct {
	Enforce     bool                `json:"enforce"`
	Domain      string              `json:"domain"`
	BindAddress string              `json:"BindAddress"`
	Channel     map[string]*Channel `json:"Channel"`
}

// GC 清理没用的数据
func (c *Config) GC() {
	for _, channel := range c.Channel {
		channel.GC()
	}
}

// Channel 通道配置
type Channel struct {
	Status    bool             `json:"status"`
	Model     string           `json:"model"`
	Secret    string           `json:"secret"`
	Issuer    string           `json:"issuer"`
	Lock      *sync.RWMutex    `json:"-"`
	Notifies  []Notify         `json:"notify"`
	Blacklist []string         `json:"blacklist"`
	Whitelist []string         `json:"whitelist"`
	Allow     map[string]int64 `json:"-"`
	Block     map[string]int64 `json:"-"`
}

// GenerateOTP 生成OTP二维码
func (c *Channel) GenerateOTP(proxy string, domain string) {
	var key, err = totp.Generate(totp.GenerateOpts{
		Period:      30,
		SecretSize:  32,
		Digits:      otp.DigitsSix,
		Algorithm:   otp.AlgorithmSHA512,
		Issuer:      domain,
		AccountName: proxy + "@" + domain,
	})

	if nil == err {
		var buf bytes.Buffer
		img, err := key.Image(200, 200)
		if err != nil {
			fmt.Print(err)
		}

		png.Encode(&buf, img)

		fmt.Printf("Issuer:       %s\n", key.Issuer())
		fmt.Printf("Account Name: %s\n", key.AccountName())
		fmt.Printf("Secret:       %s\n", key.Secret())
		fmt.Println("Writing PNG to qr-code.png....")
		ioutil.WriteFile("qr-code.png", buf.Bytes(), 0644)
		fmt.Println("")
		fmt.Println(key.String())
		fmt.Println("")
	}
}

// Check 检查一次性验证码是否有效
func (c *Channel) Check(code string) bool {
	var opt = totp.ValidateOpts{
		Period:    30,
		Skew:      1,
		Digits:    otp.DigitsSix,
		Algorithm: otp.AlgorithmSHA512,
	}
	var valid, _ = totp.ValidateCustom(code, c.Secret, time.Now().UTC(), opt)

	return valid
}

// IsAllow 是否充许访问
func (c *Channel) IsAllow(ip string) bool {
	var ts int64
	var ok, allow, block bool

	c.Lock.RLock()
	defer c.Lock.RUnlock()

	for _, v := range c.Whitelist {
		if v == ip {
			allow = true
			break
		}
	}

	if !allow {
		for _, v := range c.Blacklist {
			if v == ip {
				block = true
				break
			}
		}
	}

	if !allow && !block {
		if ts, ok = c.Block[ip]; ok {
			if time.Now().Unix() < ts {
				block = true
			}
		}
	}

	if !allow && !block {
		if ts, ok = c.Allow[ip]; ok {
			if time.Now().Unix() < ts {
				allow = true
			}
		}
	}

	return allow
}

// IsBlock 是否被屏蔽
func (c *Channel) IsBlock(ip string) (int64, bool) {
	var ts int64
	var ok, block bool

	c.Lock.RLock()
	defer c.Lock.RUnlock()

	for _, v := range c.Blacklist {
		if v == ip {
			block = true
			break
		}
	}

	if !block {
		if ts, ok = c.Block[ip]; ok {
			if time.Now().Unix()+60 < ts {
				block = true
			}
		}
	}

	return ts, block
}

// AddAllow 添加白名单
func (c *Channel) AddAllow(ip string, ts int64) {
	c.Lock.Lock()
	defer c.Lock.Unlock()

	if nil == c.Allow {
		c.Allow = make(map[string]int64, 100)
	}

	c.Allow[ip] = ts
}

// AddBlock 添加黑名单
func (c *Channel) AddBlock(ip string, ts int64, step int64) {
	c.Lock.Lock()
	defer c.Lock.Unlock()

	if nil == c.Block {
		c.Block = make(map[string]int64, 100)
	}

	if t, ok := c.Block[ip]; ok && t > ts {
		if t-ts < 60 {
			c.Block[ip] = t + step
		} else {
			c.Block[ip] = t + step*30
		}
	} else {
		c.Block[ip] = ts + step
	}
}

// GC 清理过期数据
func (c *Channel) GC() {
	var ts int64
	var ip string
	var now = time.Now().Unix()

	c.Lock.Lock()
	defer c.Lock.Unlock()

	for ip, ts = range c.Allow {
		if ts < now {
			delete(c.Allow, ip)
		}
	}
	for ip, ts = range c.Block {
		if ts < now {
			delete(c.Block, ip)
		}
	}
}

// Notify 通知配置
type Notify struct {
	Name   string                 `json:"name"`
	Action string                 `json:"action"`
	Config map[string]interface{} `json:"config"`
}
